esphome/components/i2c/i2c_bus_arduino.cpp

```diff
@@ -227,10 +230,21 @@ void ArduinoI2CBus::recover_() {
     // When SCL is kept LOW at this point, we might be looking at a device
     // that applies clock stretching. Wait for the release of the SCL line,
     // but not forever. There is no specification for the maximum allowed
-    // time. We'll stick to 500ms here.
-    auto wait = 20;
+    // time. We yield and reset the WDT, so as to avoid triggering reset.
+    // No point in trying to recover the bus by forcing a uC reset. Bus
+    // should recover in a few ms or less else not likely to recovery at
+    // all.
+    auto wait = 250;
     while (wait-- && digitalRead(scl_pin_) == LOW) {  // NOLINT
+#if defined(USE_ESP32)
+      esp_task_wdt_reset();
+      delayMicroseconds(half_period_usec * 2);
+#elif defined(USE_ESP8266)
+      EspClass::wdtFeed();
+      delayMicroseconds(half_period_usec * 2);
+#elif
       delay(25);
+#endif
```

Comment:
`App.feed_wdt()` exists as a layer so component code doesnt need to mostly check the chip.

```suggestion
      App.feed_wdt();
      delayMicroseconds(half_period_usec * 2);
```